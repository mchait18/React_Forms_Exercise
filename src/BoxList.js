import React, { useState } from "react";
import { v4 as uuid } from 'uuid';
import Box from "./Box"
import NewBoxForm from "./NewBoxForm"

const BoxList = () => {

    const [boxes, setBoxes] = useState([])

    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, { ...newBox, id: uuid() }])
    }
    const removeBox = (id) => {
        setBoxes(boxes => boxes.filter(b => b.id !== id))
    }

    return (
        <div>
            <h2>Color Box Maker</h2>
            <div>
                {boxes.map(({ id, width, height, bgColor }) =>
                    <Box
                        key={id}
                        id={id}
                        width={width}
                        height={height}
                        bgColor={bgColor}
                        removeBox={removeBox} />)}
            </div>
            <NewBoxForm addBox={addBox} />
        </div>
    )
}

export default BoxList;