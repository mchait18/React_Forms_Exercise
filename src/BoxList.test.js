import React from "react";
import { render, fireEvent, queryByLabelText } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList) {
    const heightInput = boxList.getByLabelText("Height: 0");
    const widthInput = boxList.getByLabelText("Width: 0");
    const colorInput = boxList.getByLabelText("Background Color:");
    const btn = boxList.queryByText("Add Box");
    fireEvent.change(heightInput, { target: { value: 100 } });
    fireEvent.change(widthInput, { target: { value: 100 } });
    fireEvent.change(colorInput, { target: { value: 'magenta' } });
    fireEvent.click(btn);
}
it("renders without crashing", function () {
    render(<BoxList />);
});

it("matches snapshot", function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("should add new item", function () {
    const boxList = render(<BoxList />);
    //no remove button yet
    expect(boxList.queryByText('X')).not.toBeInTheDocument();

    addBox(boxList)
    //expect to see a remove button
    expect(boxList.queryByText('X')).toBeInTheDocument();
    const removeButton = boxList.getByText('X')
    expect(removeButton.previousSibling).toHaveStyle(`
    width: 100px;
    height: 100px;
    background-color: rgb(0, 0, 0);
    `)
})
it("can remove a box", function () {
    const boxList = render(<BoxList />);
    addBox(boxList);

    const removeButton = boxList.getByText("X");

    // click the remove button and the box should be gone
    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
});