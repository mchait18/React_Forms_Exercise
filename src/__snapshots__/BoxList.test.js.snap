// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Color Box Maker
    </h2>
    <div />
    <form>
      <label
        for="height"
      >
        Height: 0
      </label>
      <input
        id="height"
        max="200"
        min="1"
        name="height"
        placeholder="Height"
        type="range"
        value="0"
      />
      <label
        for="width"
      >
        Width: 0
      </label>
      <input
        id="width"
        max="200"
        min="1"
        name="width"
        type="range"
        value="0"
      />
      <label
        for="bgColor"
      >
        Background Color: 
      </label>
      <input
        id="bgColor"
        name="bgColor"
        placeholder="Background Color"
        type="color"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
